src/main/java/fi/aalto/cs/apluscourses/intellij/model/CourseUpdater.java

```diff
@@ -0,0 +1,147 @@
+package fi.aalto.cs.apluscourses.intellij.model;
+
+import com.intellij.notification.Notification;
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.intellij.notifications.NewModulesVersionsNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.Notifier;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.Module;
+import fi.aalto.cs.apluscourses.utils.Event;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CourseUpdater {
+
+  @NotNull
+  private final Course course;
+
+  @NotNull
+  private final Project project;
+
+  @NotNull
+  private final URL courseUrl;
+
+  @NotNull
+  private final Event eventToTrigger;
+
+  @NotNull
+  private final Notifier notifier;
+
+  private final long updateInterval;
+
+  private Thread thread = null;
+
+  private final Set<String> notifiedModules = new HashSet<>();
+
+  private Notification moduleUpdatesNotification = null;
+
+  /**
+   * Construct a course updater with the given parameters.
+   */
+  public CourseUpdater(@NotNull Course course,
+                       @NotNull Project project,
+                       @NotNull URL courseUrl,
+                       @NotNull Event eventToTrigger,
+                       @NotNull Notifier notifier,
+                       long updateInterval) {
+    this.course = course;
+    this.project = project;
+    this.courseUrl = courseUrl;
+    this.eventToTrigger = eventToTrigger;
+    this.notifier = notifier;
+    this.updateInterval = updateInterval;
+  }
+
+  /**
+   * Starts or restarts this updater. This method is thread safe and can be called from multiple
+   * threads.
+   */
+  public synchronized void restart() {
+    if (thread != null) {
+      thread.interrupt();
```

Comment:
Would it be safest to join the old thread before continuing? Surely, the new thread cannot do anything before the old one releases the lock but still, it could help reasoning things out. Like, then it would be guaranteed that after `restart` returns, the old thread cannot trigger events, and so on.

On the other hand, it would make this method more prone to unnoticed concurrency bugs causing deadlock.

Reply:
Interesting idea! That would indeed give the same guarantees as the run lock mechanism, and it would also make it very clear that the previous thread is completely done with everything. It would actually render the current run lock unnecessary I think.

The main difference with the run lock would be that with a `join`, the `restart` method (and by extension its caller) now blocks until the running thread has reacted to the interruption. With the current locking mechanism, only the newly created thread is the one that blocks until the old thread is done.

`restart` blocking until the thread has reacted to the interruption makes me a bit nervous, because we fetch network resources in the update loop, which could cause unpredictable delays. Apache's HTTP library surely has some reasonable default timeouts, but I haven't looked into it